CheckAzurePipelinesTestResults.ValidateExpectedTestFailuresAsync needs to handle AzDO errors better

From: https://dev.azure.com/dnceng/public/_build/results?buildId=387506&view=logs&s=859b8d9a-8fd6-5a5c-6f5e-f84f1990894e&j=5fae9d80-faa2-578a-a809-a74b747965c3

```
Object reference not set to an instance of an object. [D:\a\1\s\tests\UnitTests.proj]
   at Microsoft.DotNet.Helix.AzureDevOps.CheckAzurePipelinesTestResults.ValidateExpectedTestFailuresAsync(HttpClient client) in /_/src/Microsoft.DotNet.Helix/Sdk/CheckAzurePipelinesTestResults.cs:line 130 [D:\a\1\s\tests\UnitTests.proj]
   at Microsoft.DotNet.Helix.AzureDevOps.CheckAzurePipelinesTestResults.ExecuteCoreAsync(HttpClient client) in /_/src/Microsoft.DotNet.Helix/Sdk/CheckAzurePipelinesTestResults.cs:line 23 [D:\a\1\s\tests\UnitTests.proj]
   at Microsoft.DotNet.Helix.AzureDevOps.AzureDevOpsTask.Execute() in /_/src/Microsoft.DotNet.Helix/Sdk/AzureDevOpsTask.cs:line 67 [D:\a\1\s\tests\UnitTests.proj]
##[error]src\Microsoft.DotNet.Helix\Sdk\tools\azure-pipelines\AzurePipelines.MultiQueue.targets(43,5): error : Object reference not set to an instance of an object.
```
This occurred during an Azure DevOps outage, most likely because it attempted to deserialize some JSON and actually got HTML.  We should fail in a more friendly way here, ideally with some retries as well (in this case, it would likely have still failed as the outage was several minutes)

--- a/src/Microsoft.DotNet.Helix/Sdk/AzureDevOpsTask.cs
+++ b/src/Microsoft.DotNet.Helix/Sdk/AzureDevOpsTask.cs
@@ -96,12 +96,12 @@ namespace Microsoft.DotNet.Helix.AzureDevOps
             }
         }
 
-        protected Task<T> RetryAsync<T>(Func<Task<T>> function)
+        protected async Task<T> RetryAsync<T>(Func<Task<T>> function)
         {
             // Grab the retry logic from the helix api client
             try
             {
-                return ApiFactory.GetAnonymous().RetryAsync(
+                return await ApiFactory.GetAnonymous().RetryAsync(
                         async () => await function(),
                         ex => Log.LogMessage(MessageImportance.Low, $"Azure Dev Ops Operation failed: {ex}\nRetrying..."),
                         CancellationToken.None);
@@ -109,7 +109,7 @@ namespace Microsoft.DotNet.Helix.AzureDevOps
             catch (HttpRequestException ex)
             {
                 Log.LogError(FailureCategory.Helix, ex.ToString());
-                return null;
+                return default;
             }
         }
 

--- a/src/Microsoft.DotNet.Helix/Sdk/AzureDevOpsTask.cs
+++ b/src/Microsoft.DotNet.Helix/Sdk/AzureDevOpsTask.cs
@@ -76,43 +76,65 @@ namespace Microsoft.DotNet.Helix.AzureDevOps
             return !Log.HasLoggedErrors;
         }
 
-        protected Task RetryAsync(Func<Task> function)
+        protected async Task RetryAsync(Func<Task> function)
         {
-            // Grab the retry logic from the helix api client
-            return ApiFactory.GetAnonymous().RetryAsync(
-                    async () =>
-                    {
-                        await function();
-                        return false; // the retry function requires a return, give it one
-                    },
-                    ex => Log.LogMessage(MessageImportance.Low, $"Azure Dev Ops Operation failed: {ex}\nRetrying..."),
-                    CancellationToken.None);
-
+            try
+            {
+                // Grab the retry logic from the helix api client
+                await ApiFactory.GetAnonymous().RetryAsync(
+                        async () =>
+                        {
+                            await function();
+                            return false; // the retry function requires a return, give it one
+                        },
+                        ex => Log.LogMessage(MessageImportance.Low, $"Azure Dev Ops Operation failed: {ex}\nRetrying..."),
+                        CancellationToken.None);
+            }
+            catch (HttpRequestException ex)
+            {
+                Log.LogError(FailureCategory.Helix, ex.ToString());
+            }
         }
 
         protected Task<T> RetryAsync<T>(Func<Task<T>> function)
         {
             // Grab the retry logic from the helix api client
-            return ApiFactory.GetAnonymous().RetryAsync(
-                    async () => await function(),
-                    ex => Log.LogMessage(MessageImportance.Low, $"Azure Dev Ops Operation failed: {ex}\nRetrying..."),
-                    CancellationToken.None);
-
+            try
+            {
+                return ApiFactory.GetAnonymous().RetryAsync(
+                        async () => await function(),
+                        ex => Log.LogMessage(MessageImportance.Low, $"Azure Dev Ops Operation failed: {ex}\nRetrying..."),
+                        CancellationToken.None);
+            }
+            catch (HttpRequestException ex)
+            {
+                Log.LogError(FailureCategory.Helix, ex.ToString());
+                return null;
+            }
         }
 
-        protected async Task LogFailedRequest(HttpRequestMessage req, HttpResponseMessage res)
+        protected async Task HandleFailedRequest(HttpRequestMessage req, HttpResponseMessage res)
         {
-            int statusCodeValue = (int) res.StatusCode;
-            FailureCategory category = statusCodeValue >= 400 && statusCodeValue < 500
-                ? FailureCategory.Build
-                : FailureCategory.Helix;
-
-            Log.LogError(category, $"Request to {req.RequestUri} returned failed status {statusCodeValue} {res.ReasonPhrase}\n\n{(res.Content != null ? await res.Content.ReadAsStringAsync() : "")}");
             if (res.StatusCode == HttpStatusCode.Found)
             {
                 Log.LogError(
                     FailureCategory.Build,
                     "A call to an Azure DevOps api returned 302 Indicating a bad 'System.AccessToken' value.\n\nPlease Check the 'Make secrets available to builds of forks' in the pipeline pull request validation trigger settings.\nWe have evaluated the security considerations of this setting and have determined that it is fine to use for our public PR validation builds.");
+                return;
+            }
+
+            var statusCodeValue = (int)res.StatusCode;
+            var message = $"Request to {req.RequestUri} returned failed status {statusCodeValue} {res.ReasonPhrase}\n\n{(res.Content != null ? await res.Content.ReadAsStringAsync() : "")}";
+
+            if (statusCodeValue >= 400 && statusCodeValue < 500)
+            {
+                Log.LogError(FailureCategory.Build, message);
+            }
+            else
+            {
+                // we want to engage retry logic from HelixApi.RetryAsync in this case
+                Log.LogMessage(MessageImportance.Normal, message);
+                throw new HttpRequestException(message);
             }
         }
 
@@ -120,7 +142,7 @@ namespace Microsoft.DotNet.Helix.AzureDevOps
         {
             if (!res.IsSuccessStatusCode)
             {
-                await LogFailedRequest(req, res);
+                await HandleFailedRequest(req, res);
                 return null;
             }
 

--- a/src/Microsoft.DotNet.Helix/Sdk/CheckAzurePipelinesTestResults.cs
+++ b/src/Microsoft.DotNet.Helix/Sdk/CheckAzurePipelinesTestResults.cs
@@ -52,6 +52,7 @@ namespace Microsoft.DotNet.Helix.AzureDevOps
                             }
                         }
                     });
+
                 if (data != null && data["runStatistics"] is JArray runStatistics)
                 {
                     var failed = runStatistics.Children()
@@ -84,6 +85,7 @@ namespace Microsoft.DotNet.Helix.AzureDevOps
                         }
                     }
                 });
+
             if (data != null && data["aggregatedResultsAnalysis"] is JObject aggregatedResultsAnalysis &&
                 aggregatedResultsAnalysis["resultsByOutcome"] is JObject resultsByOutcome)
             {
@@ -128,25 +130,28 @@ namespace Microsoft.DotNet.Helix.AzureDevOps
                         }
                     });
 
-                var failedResults = (JArray) data["value"];
-                HashSet<string> expectedFailures = ExpectedTestFailures?.Select(i => i.GetMetadata("Identity")).ToHashSet() ?? new HashSet<string>();
-                foreach (var failedResult in failedResults)
+                if (data != null)
                 {
-                    var testName = (string) failedResult["automatedTestName"];
-                    if (expectedFailures.Contains(testName))
+                    var failedResults = (JArray)data["value"];
+                    HashSet<string> expectedFailures = ExpectedTestFailures?.Select(i => i.GetMetadata("Identity")).ToHashSet() ?? new HashSet<string>();
+                    foreach (var failedResult in failedResults)
                     {
-                        expectedFailures.Remove(testName);
-                        Log.LogMessage($"TestRun {runId}: Test {testName} has failed and was expected to fail.");
+                        var testName = (string)failedResult["automatedTestName"];
+                        if (expectedFailures.Contains(testName))
+                        {
+                            expectedFailures.Remove(testName);
+                            Log.LogMessage($"TestRun {runId}: Test {testName} has failed and was expected to fail.");
+                        }
+                        else
+                        {
+                            Log.LogError(FailureCategory.Test, $"TestRun {runId}: Test {testName} has failed and is not expected to fail.");
+                        }
                     }
-                    else
-                    {
-                        Log.LogError(FailureCategory.Test, $"TestRun {runId}: Test {testName} has failed and is not expected to fail.");
-                    }
-                }
 
-                foreach (string expectedFailure in expectedFailures)
-                {
-                    Log.LogError(FailureCategory.Test, $"TestRun {runId}: Test {expectedFailure} was expected to fail but did not fail.");
+                    foreach (string expectedFailure in expectedFailures)
+                    {
+                        Log.LogError(FailureCategory.Test, $"TestRun {runId}: Test {expectedFailure} was expected to fail but did not fail.");
+                    }
                 }
             }
         }

